tests/test_soil_carbon.py

```diff
@@ -61,69 +61,233 @@ def test_soil_carbon_class(caplog, maom, lmwc, raises, expected_log_entries):
-    pH = np.array([7.0, 7.0], dtype=np.float32)
-    bulk_density = np.array([1350, 1350], dtype=np.float32)
-    percent_clay = np.array([50.0, 50.0], dtype=np.float32)
-    soil_moisture = np.array([0.5, 0.5], dtype=np.float32)
-    soil_temp = np.array([35.0, 35.0], dtype=np.float32)
-    dt = 2.0 / 24.0
-
-    soil_carbon.update_pools(
-        pH, bulk_density, soil_moisture, soil_temp, percent_clay, dt
+    soil_carbon = SoilCarbonPools(
+        maom=np.array(maom, dtype=np.float32), lmwc=np.array(lmwc, dtype=np.float32)
+    )
+
+    # Mock required values
+    mock_l_to_m = mocker.patch(
+        "virtual_rainforest.models.soil.carbon.SoilCarbonPools.mineral_association"
     )
+    mock_l_to_m.return_value = np.array(lmwc_to_maom, dtype=np.float32)
+
+    soil_carbon.update_pools([], [], [], [], [], dt)
 
     # Check that pools are correctly incremented
-    assert np.allclose(soil_carbon.maom, np.array([28.8263, 25.7322]))
-    assert np.allclose(soil_carbon.lmwc, np.array([92.1736, 52.2677]))
+    assert np.allclose(soil_carbon.maom, np.array(end_maom))
+    assert np.allclose(soil_carbon.lmwc, np.array(end_lmwc))
 
 
-def test_mineral_association():
+@pytest.mark.parametrize(
+    "maom,lmwc,temp_scalar,moist_scalar,equib_maom,Q_max,output_l_to_m",
+    [
+        (
+            [2.5, 1.7],
+            [0.05, 0.02],
+            [1.27113, 1.27196],
+            [0.750035, 0.947787],
+            [19900.19, 969.4813],
+            [2.385207e6, 1.980259e6],
+            [0.000397665, 1.178336e-5],
+        ),
+        ([4.5], [0.1], [1.27263], [0.880671], [832.6088], [647142.61], [0.0001434178]),
+        ([0.5], [0.005], [1.26344], [0.167814], [742.4128], [2.805371e6], [2.80359e-7]),
+    ],
+)
+def test_mineral_association(
+    mocker, maom, lmwc, temp_scalar, moist_scalar, equib_maom, Q_max, output_l_to_m
+):
     """Test that mineral_association runs and generates the correct values."""
 
     # Initialise soil carbon class
-    maom = np.array([23.0, 23.0], dtype=np.float32)
-    lmwc = np.array([98.0, 55.0], dtype=np.float32)
-    soil_carbon = SoilCarbonPools(maom, lmwc)
-
-    # Define all the required variables to run function
-    pH = np.array([7.0, 7.0], dtype=np.float32)
-    bulk_density = np.array([1350, 1350], dtype=np.float32)
-    percent_clay = np.array([50.0, 50.0], dtype=np.float32)
-    soil_moisture = np.array([0.5, 0.5], dtype=np.float32)
-    soil_temp = np.array([35.0, 35.0], dtype=np.float32)
-
-    lmwc_to_maom = soil_carbon.mineral_association(
-        pH, bulk_density, soil_moisture, soil_temp, percent_clay
+    soil_carbon = SoilCarbonPools(
+        maom=np.array(maom, dtype=np.float32), lmwc=np.array(lmwc, dtype=np.float32)
+    )
+
+    # Mock required values
+    mock_t_scalar = mocker.patch(
+        "virtual_rainforest.models.soil.carbon.convert_temperature_to_scalar"
+    )
+    mock_t_scalar.return_value = np.array(temp_scalar, dtype=np.float32)
```

Comment:
Well, yes it does isolate the problem but mocking absolutely everything but the function you are testing:

* is a lot of extra typing,
*  makes the test flow much less clear,
*  and is really difficult to debug if there is a typo in your mock value because the source of that value is obscure.

Sure, if a low level function breaks, you _will_ get a cascade of dependent failures, but that dependency is real and I think relying on units that are tested elsewhere is part of the test framework. And is _much_ easier to write.